plenum/server/replica.py

```diff
@@ -1034,26 +1042,62 @@ def process_three_phase_msg(self, msg: ThreePhaseMsg, sender: str):
             self.stasher.stash((msg, sender), result)
 
     def _process_valid_preprepare(self, pre_prepare, sender):
-        # TODO: rename to apply_pre_prepare
-        if not self.node.isParticipating:
-            self.discard(pre_prepare, 'node is not participating', self.logger.debug)
-            return None
-        pre_state_root = self.stateRootHash(pre_prepare.ledgerId, to_str=False)
-        why_not_applied = self._apply_pre_prepare(pre_prepare, sender)
+        old_state_root = self.stateRootHash(pre_prepare.ledgerId, to_str=False)
+        old_txn_root = self.txnRootHash(pre_prepare.ledgerId)
+        if self.isMaster:
+            self.logger.debug('{} state root before processing {} is {}, {}'.format(
+                self,
+                pre_prepare,
+                old_state_root,
+                old_txn_root))
+
+        # 1. APPLY
+        reqs, invalid_indices, rejects = self._apply_pre_prepare(pre_prepare)
+
+        # 2. CHECK IF MORE CHUNKS NEED TO BE APPLIED FURTHER BEFORE VALIDATION
+        if pre_prepare.sub_seq_no != 0:
+            return PP_SUB_SEQ_NO_WRONG
+
+        if not pre_prepare.final:
+            return PP_NOT_FINAL
+
+        # 3. VALIDATE APPLIED
+        invalid_from_pp = invalid_index_serializer.deserialize(pre_prepare.discarded)
+        why_not_applied = self._validate_applied_pre_prepare(pre_prepare,
+                                                             reqs, invalid_indices, invalid_from_pp)
```

Comment:
Hooray, `_apply_pre_prepare` monstrosity is finally split into smaller chunks! Thanks a lot! :tada: